semantic-analysis/src/Analysis/Typecheck.hs

```diff
@@ -21,154 +22,144 @@ module Analysis.Typecheck
+  = Var a
   | Unit
+  | Bool
+  | Int
   | String
-  | f a :-> f a
-  | Record (Map.Map Name (f a))
-  deriving (Foldable, Functor, Generic1, Traversable)
+  | Monotype a :-> Monotype a
+  -- | (Locally) undefined names whose types are unknown. May not be eliminated by unification.
+  | TypeOf Name
+  deriving (Eq, Foldable, Functor, Generic1, Ord, Show, Traversable)
 
 infixr 0 :->
 
-type Type = Term Monotype Meta
-
-type Addr = Name
-
 -- FIXME: Union the effects/annotations on the operands.
 
 -- | We derive the 'Semigroup' instance for types to take the second argument. This is equivalent to stating that the type of an imperative sequence of statements is the type of its final statement.
-deriving via (Last (Term Monotype a)) instance Semigroup (Term Monotype a)
+deriving via (Last (Monotype a)) instance Semigroup (Monotype a)
+
+instance Applicative Monotype where
+  pure = Var
+  (<*>) = ap
+
+instance Monad Monotype where
+  m >>= f = case m of
+    Var a    -> f a
+    Unit     -> Unit
+    Bool     -> Bool
+    Int      -> Int
+    String   -> String
+    TypeOf n -> TypeOf n
+    a :-> b  -> (a >>= f) :-> (b >>= f)
+
 
-deriving instance (Eq   a, forall a . Eq   a => Eq   (f a), Monad f) => Eq   (Monotype f a)
-deriving instance (Ord  a, forall a . Eq   a => Eq   (f a)
-                         , forall a . Ord  a => Ord  (f a), Monad f) => Ord  (Monotype f a)
-deriving instance (Show a, forall a . Show a => Show (f a))          => Show (Monotype f a)
+type Type = Monotype Meta
 
-instance HFunctor Monotype
-instance RightModule Monotype where
-  Unit      >>=* _ = Unit
-  Bool      >>=* _ = Bool
-  String    >>=* _ = String
-  (a :-> b) >>=* f = a >>= f :-> b >>= f
-  Record m  >>=* f = Record ((>>= f) <$> m)
 
 type Meta = Int
 
-newtype Polytype f a = PForAll (Scope () f a)
-  deriving (Foldable, Functor, Generic1, HFunctor, RightModule, Traversable)
 
-deriving instance (Eq   a, forall a . Eq   a => Eq   (f a), Monad f) => Eq   (Polytype f a)
-deriving instance (Ord  a, forall a . Eq   a => Eq   (f a)
-                         , forall a . Ord  a => Ord  (f a), Monad f) => Ord  (Polytype f a)
-deriving instance (Show a, forall a . Show a => Show (f a))          => Show (Polytype f a)
+data Polytype a
+  = PForAll (Polytype (Maybe a))
+  | PType (Monotype a)
+  deriving (Eq, Foldable, Functor, Generic1, Ord, Show, Traversable)
 
 
-forAll :: (Eq a, Syntax.Has Polytype sig m) => a -> m a -> m a
-forAll n body = Syntax.send (PForAll (abstract1 n body))
+closed :: Traversable f => f a -> Maybe (f b)
+closed = traverse (const Nothing)
 
-forAlls :: (Eq a, Syntax.Has Polytype sig m, Foldable t) => t a -> m a -> m a
+abstract :: Eq a => a -> Polytype a -> Polytype (Maybe a)
+abstract n = fmap (\ a -> a <$ guard (a == n))
+
+
+forAll :: Eq a => a -> Polytype a -> Polytype a
+forAll n body = PForAll (abstract n body)
+
+forAlls :: (Eq a, Foldable t) => t a -> Polytype a -> Polytype a
 forAlls ns body = foldr forAll body ns
```

Comment:
This is the _I Am Not a Number, I Am a Free Variable_ approach. It doesn’t give the fastest abstraction and substitution (relative to e.g. `bound`, but a) that doesn’t actually matter at all for our purposes, and b) the smarter next move would be HOAS anyway.